src/Text-Core/RunArray.class.st

```diff
@@ -153,90 +153,57 @@ RunArray >> , aRunArray [
-			aRunArray do: [:each | new addLast: each].
-			^new].
-	runs size = 0 ifTrue: [^aRunArray copy].
-	aRunArray runs size = 0 ifTrue: [^self copy].
-	(values at: values size) ~= (aRunArray values at: 1)
-		ifTrue: [^RunArray
-					runs: runs , aRunArray runs
-					values: values , aRunArray values].
-	newRuns := runs
-			copyReplaceFrom: runs size
-			to: runs size
-			with: aRunArray runs.
-	newRuns at: runs size put: (runs at: runs size) + (aRunArray runs at: 1).
-	^RunArray
-		runs: newRuns
-		values:
-			(values
-				copyReplaceFrom: values size
-				to: values size
-				with: aRunArray values)
+	| newRuns |
+	aRunArray class = self class ifFalse: [
+		| new |
+		new := self copy.
+		"attempt to be sociable"
+		new addAll: aRunArray.
+		^ new ].
+	runs ifEmpty: [ ^ aRunArray copy ].
+	aRunArray runs ifEmpty: [ ^ self copy ].
+	values last ~= aRunArray values first ifTrue: [ ^ self class runs: runs , aRunArray runs values: values , aRunArray values ].
+	newRuns := runs copyReplaceFrom: runs size to: runs size with: aRunArray runs.
+	newRuns at: runs size put: runs last + aRunArray runs first.
+	^ self class runs: newRuns values: (values copyReplaceFrom: values size to: values size with: aRunArray values)
 ]
 
 { #category : #comparing }
 RunArray >> = otherArray [
 	"Test if all my elements are equal to those of otherArray"
 
-	self == otherArray ifTrue: [^true].
-	self species == otherArray species ifFalse: [^ false].
-	(otherArray isMemberOf: RunArray) ifFalse: [^ self hasEqualElements: otherArray].
+	self == otherArray ifTrue: [ ^ true ].
+	otherArray class = self class ifFalse: [ ^ self hasEqualElements: otherArray ].
+	self species == otherArray species ifFalse: [ ^ false ].
 
 	"Faster test between two RunArrays"
-	^ (runs hasEqualElements: otherArray runs)
-		and: [values hasEqualElements: otherArray values]
+	^ (runs hasEqualElements: otherArray runs) and: [ values hasEqualElements: otherArray values ]
 ]
 
 { #category : #adding }
-RunArray >> add: newObject [
-	"Include newObject as one of the receiver's elements. Answer newObject.
-	ArrayedCollections cannot respond to this message."
+RunArray >> add: value [
+	"Add value as the last element of the receiver."
 
-	^ self addLast: newObject
+	lastIndex := nil. "flush access cache"
+	(runs isEmpty or: [ values last ~= value ])
+		ifTrue: [
+			runs := runs copyWith: 1.
+			values := values copyWith: value ]
+		ifFalse: [ runs at: runs size put: runs last + 1 ].
+	^ value
 ]
 
 { #category : #adding }
 RunArray >> add: newObject withOccurrences: anInteger [
 	"Add value as the last element of the receiver, the given number of times"
-	anInteger <= 0 ifTrue: [ ^self ].
-	lastIndex := nil. "flush access cache"
-	(runs size=0 or: [values last ~= newObject])
-		ifTrue:
-			[runs := runs copyWith: anInteger.
-			values := values copyWith: newObject]
-		ifFalse:
-			[runs at: runs size put: runs last + anInteger]
-]
-
-{ #category : #adding }
-RunArray >> addFirst: value [
```

Comment:
RunArray could be used independently of Text. There is no reason to limit usage, nor anything in the comment of the class that refer to Test

Reply:
If that is the case maybe it should not be packaged in Text-Core but in a Collection package?

Reply:
Packages seems to me weird and unreliable, I think that because there is no namespace and because dependency resolution is mostly faith-based, people do not care that much.

Where do you store a type-agnostic data structure but with a specific use case (long sequences of equal elements) and a single main client in the core classes? There is no good answer here :( so status quo might win.

Reply:
Personally, I do not use classes that seems to be part of something unrelated to me. 

Having a Collection implemented in Text-Core make it a Text collection to me that should not be used outside the Text management. 

To have something clear here, I think that we should either packages it in a general package like Collections-Sequenceable or assume it's specific to text and remove code that is unspecific to tests. 

After having this discussion, I would opt to move it to Collections-Sequenceable.

Reply:
Seriously, I do not know. Maybe it's out of scope of the current PR.
But it's possible that some people are using RunArray (it even has a general non-prefixed public name, it's an array that runs!) so I would put the method back in to be safe. They are not that big

Reply:
I agrée that this out of the scope of this PR. I’m just wondering how things should be ideally :)